### DISCUSSION TITLE
Question on Chap. 5.1: AbstractBlock.getOutputShapes

### DISCUSSION THREAD

**Discussion by hmf** [deepjavalibrary/djl#752] [Q&A] [answered]
I am going through the examples in chapter 5.1 and coding these in Scala3. I don't seem to understand the `AbstractBlock.getOutputShapes` method. 

First the method name seems to imply that we return several shapes, however we only return one - the last one. Is this correct? 

Second, the code `block.getOutputShapes` seems to be taking the information of the list of ordered blocks set up in `initializeChildBlocks` and *simply* returns the blocks output shape. 
Is this correct? Are any additional checks made by the block? Why do we need to override this method? It seems that this would be the default behavior.

BTW, with the changes in the latest update the example's `forward` method should be renamed `forwardInternal`. Am I correct?

TIA.

**Comment by zachgk** [comment, marked as answer]
The getOutputShapes should return what the output NDArray shapes would be, for a given input into your custom Block. It uses an array of shapes as an input and output because blocks use NDLists, not NDArrays. So, each element in the array of shapes is the shape for a different NDArray in the NDList. If you only pass a singleton NDList (as is quite common), you would have arrays containing only one shape.

The getOutputShapes is about the block's forward method. Knowing what shapes the forward will input or output is used for both printing blocks and initialization. So, initializeChildBlocks might be using this information to help determine the inputs for various child blocks.

Yeah, we added some additional bookkeeping and initialization logic to forward. So, the block implementer will be overriding the forwardInternal which comes after the bookkeeping now

**Reply by hmf** [reply]
> The getOutputShapes should return what the output NDArray shapes would be, for a given input into your custom Block.

Think I have it now. This function can: a) query child blocks for the information, b) return a constant shape irrespective of input shape or determine the output based on the input. 

In the last case I assume that if I have a block with output Shape(1,n) but get an input Size(m,n) then the calculate output would be Size(m,n) - batch size m. I am also assuming that if I use a Translator, the input Shape is determined by the translator. 
 
If this is incorrect, please correct me.

Thank you
